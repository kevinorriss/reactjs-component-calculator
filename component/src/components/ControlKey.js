import React from 'react'
import PropTypes from 'prop-types'

class ControlKey extends React.Component {
    constructor(props) {
        super(props)
        this.handleKeyDown = this.handleKeyDown.bind(this)
        this.state = {
            text: props.text,
            keys: props.keys,
            onClick: props.onClick,
            className: props.className
        }
    }

    componentDidMount() {
        window.addEventListener('keydown', this.handleKeyDown)
    }
    
    componentWillUnmount() {
        window.removeEventListener('keydown', this.handleKeyDown)
    }

    handleKeyDown(event) {
        if (this.state.keys.includes(event.key)) {
            this.state.onClick()
        }
    }

    render() {
        return (
            <div className={"button button--control" + (this.state.className ? ' ' + this.state.className : '')} onClick={() => {this.state.onClick()}}>
                <div>{this.state.text}</div>
            </div>
        )
    }
}

ControlKey.propTypes = {
    text: PropTypes.string.isRequired,
    keys: PropTypes.array.isRequired,
    onClick: PropTypes.func.isRequired,
    className: PropTypes.string
}

export default ControlKey