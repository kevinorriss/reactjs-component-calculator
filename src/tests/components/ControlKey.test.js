import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import ControlKey from '../../../component/src/components/ControlKey'

it('should render ControlKey correctly', () => {
    const component = renderer.create(<ControlKey text="C" keys={['Escape']} onClick={jest.fn()} className="button--control-equals" />).toJSON()
    expect(component).toMatchSnapshot()
})

it('should call onClick when corresponding keyboard key pressed', () => {
    const onCick = jest.fn()
    const map = {}
    window.addEventListener = jest.fn((event, cb) => { map[event] = cb })
    shallow(<ControlKey text="C" keys={['Escape']} onClick={onCick}/>)
    map.keydown({ key: 'Escape' })
    expect(onCick).toHaveBeenCalled()
})

it('should call onClick when clicked', () => {
    const onCick = jest.fn()
    const wrapper = shallow(<ControlKey text="C" keys={['Escape']} onClick={onCick}/>)
    wrapper.find('.button--control').simulate('click')
    expect(onCick).toHaveBeenCalled()
})