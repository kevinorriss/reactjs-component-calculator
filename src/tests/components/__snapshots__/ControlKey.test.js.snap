// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ControlKey correctly 1`] = `
<div
  className="button button--control button--control-equals"
  onClick={[Function]}
>
  <div>
    C
  </div>
</div>
`;
